Make cproverNondetInitialize be inherited [TG-8677]

owen-mc-diffblue opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

owen-mc-diffblue committed:
Add tests for inheritance of cproverNondetInitialize

owen-mc-diffblue committed:
Inherit cproverNondetInitialize like a virtual method

We should look for it on the super class, and on that's super class,
and so on till we reach the top of the class hierarchy.

owen-mc-diffblue committed:
Fix unit test to set up symbol table correctly

It didn't add proper type for java.lang.Object symbol, which
caused find_cprover_nondet_initialise to fall over.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4097d449e6850238f01678804a3a14871ed88886).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/127175289

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 62b5999be9a0501c4c18263866571007e806bfcc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/127332523

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 38a93c44582eae8ef1e590201c7e128ec2319a68).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/127404727

JohnDumbell commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -510,3 +511,41 @@ void set_declaring_class(symbolt &symbol, const irep_idt &declaring_class)
+  return nullptr;
+}
+
+const symbolt *find_cprover_nondet_initialize(
```
Is `optionalt` usable here? I know it causes problems in certain situations (sadly), so wouldn't be surprised if the answer is no.

owen-mc-diffblue replied:
You can't have an `optionalt` of a reference. They say to use a pointer. I could use optionalt of a reference wrapper, I think.

JohnDumbell commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -510,3 +511,41 @@ void set_declaring_class(symbolt &symbol, const irep_idt &declaring_class)
+  while(!ret)
+  {
+    const irep_idt &class_id =
+      to_java_class_type(ns.lookup(class_name).type).get_super_class();
```
Worth being more defensive here? I can't remember if `ns.lookup` throws on not found or returns a default. Same with the call on `get_super_class` if the type is dodgy.

owen-mc-diffblue replied:
`ns.lookup` has an INVARIANT that the thing was found. That's the only reason I'm using a namespace instead of the symbol table directly, to avoid having to handle that error myself.

`to_java_class_type` asserts the type has ID_struct. If it doesn't have a super class then `get_super_class()` will return an empty irep_idt, which is handled on the next line.

So actually I think the errors are handled okay, despite how little explicit error handling there is.

thk123 commented on jbmc/regression/jbmc-inheritance/cprover_nondet_initialize/inheritanceFromAbstractClass.desc:
```diff
@@ -0,0 +1,10 @@
+Test.class
+--function Test.inheritanceFromAbstractClass --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
+^EXIT=0$
+^SIGNAL=0$
```
:no_entry_sign: please add more asserts (e.g. a verification successful) :rescue_worker_helmet:

thk123 commented on jbmc/regression/jbmc-inheritance/cprover_nondet_initialize/Test.java:
```diff
@@ -0,0 +1,45 @@
+
+  public void cproverNondetInitialize() { CProver.assume(x == 42); }
+
+  abstract void neverCalled();
```
:question: what is the purpose of this method?

owen-mc-diffblue replied:
I thought an abstract class needed an abstract method. I'll remove it.

thk123 commented on jbmc/regression/jbmc-inheritance/cprover_nondet_initialize/inheritanceTwoLevels.desc:
```diff
@@ -0,0 +1,10 @@
+Test.class
+--function Test.inheritanceTwoLevels --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
+^EXIT=0$
+^SIGNAL=0$
```
:rescue_worker_helmet:

thk123 commented on scripts/pretty-printers/gdb/pretty_printers.py:
```diff
@@ -1,4 +1,21 @@
 import gdb
+import json
```
:no_entry_sign: please don't put prettty printing changes into the same commit as the body of the work

thk123 replied:
Or even in this PR??

thk123 replied:
I've not reviewed these changes on the assumption they are in error!

owen-mc-diffblue replied:
Oops

thk123 commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -510,3 +511,47 @@ void set_declaring_class(symbolt &symbol, const irep_idt &declaring_class)
+  return nullptr;
+}
+
+const symbolt *find_cprover_nondet_initialize(
```
:bulb: great candidate for a unit test

thk123 reviewed (commented):
:no_entry_sign: please add a test for class that implements an interface :heavy_check_mark: 
:no_entry_sign: please add a test for calling a base cproverNondetInitalize from derived :heavy_check_mark: 
:question: does virtual function resolution make sense to test? answered IRL - no
:question: can an interface define `cproverNondetInitalize` - answer IRL - we don't yet support default methods so don't need to worry about this

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3b87e83685a6f9059af3972548da929f57f1aecb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128040215

owen-mc-diffblue commented:
@thk123 I've address most of the review comments. I haven't added tests for interfaces because I don't know how they would implement cproverNondetInitialize (it seems it is possible using default implementations, which came in in java 8, but I'm not sure we support that because I've never heard anyone mention it).

The one thing I think it remaining is unit testing `find_cprover_nondet_initialize`. @thomasspriggs suggested using `resolve_inherited_componentt` instead, which needs a class hierarchy. I have switched to using this in one of the two places, where it is easy to get one. The other place is in the java object factory. I don't think it is worth making every user of the java object factory supply a populated class hierarchy. I think it would be too slow building it each time. I think the best solution is moving my code into `resolve_inherited_component.cpp`, using the same interface, and making it an alternative implementation that doesn't need a class hierarchy.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 366d3eccec53342e799c57b1f0a4f159636cdddb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128291529

thk123 commented:
I still think a test with an interface just to ensure you're going up the correct side of the inheritance tree has value (since not reusing the `resolve_inherited_compnoentt`) but other than that lgtm

owen-mc-diffblue commented:
With https://github.com/diffblue/cbmc/pull/5118 I should be able to use `resolve_inherited_componentt`, and hence I can get rid of my little utility functions, and even better I won't have to write unit tests for them :smile:

owen-mc-diffblue commented:
@thk123 I've now merged #5118 and rebased this on top of it, so I can use `resolve_inherited_componentt` in both places. That means I no longer have the utility function which you wanted me to unit test. (I didn't do this just to avoid unit testing them - it's also much nicer code now.) I've also added a test that we find cproverNondetInitialize on a super class even when the base class also implements an interface.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 13eac315352160e55519255e76697fbe1b5178fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128470565

thk123 reviewed (approved):
Blocking comments addressed
